### ISSUE TITLE
Sync latest upstream

### ISSUE THREAD

**Issue by zpao** [facebook/esprima#78]
I started to do it but there were more than a couple conflicts and I haven't been following closely enough to fix them. I'll leave that to the experts (cc @gabelevi & @jeffmo).

There aren't too many changes: https://github.com/ariya/esprima/compare/8d62912...harmony

**Comment by mathiasbynens** [comment]
FWIW, this is blocking https://github.com/mathiasbynens/regexpu/issues/12.

**Comment by danez** [comment]
They have also landed some commits that fix tokenizing regular expressions, which are critical for jscs when used with esprima-fb.
https://github.com/ariya/esprima/commit/fc79b9d827e4c5549ed845e5ecb27f2014e764c0
see jscs-dev/node-jscs#685

Please sync.

**Comment by jeffmo** [comment]
There is some non-trivial work to be done to merge now that we've changed several aspects of our fork, so I went ahead and pushed `9001.1.0` which is what we have merged right now (to unblock those that need that). I will begin merging in the latest changes from upstream later tonight and hopefully get that pushed out later tonight if (barring any unexpectedly difficult merge work).

We're in the middle of our peer review cycle right now, so people are pretty distracted today and tomorrow getting all their last minute stuff in. I know I've dropped the ball here for several weeks now, but just wanted to give an update and will do my best to get this out ASAP.
